lib/manager/terraform/lockfile/util.ts

```diff
@@ -0,0 +1,204 @@
+      constraints,
+      hashes,
+      lineNumbers: relativeLineNumbers,
+    };
+    return lock;
+  });
+
+  if (locks.length === 0) {
+    return null;
+  }
+  return locks;
+}
+
+export function isPinnedVersion(value: string): boolean {
+  const versioning = getVersioning('hashicorp');
+  return !!versioning.isSingleVersion(value);
+}
+
+export function writeLockUpdates(
+  updates: ProviderLockUpdate[],
+  oldLockFileContent: string
+): Promise<UpdateArtifactsResult[]> {
+  const lines = oldLockFileContent.split('\n');
+
+  const sections: string[][] = [];
+  updates.forEach((update, index, array) => {
+    // re add leading whitespace
+    let startWhitespace;
+    if (index > 0) {
+      // get end of the
+      startWhitespace = array[index - 1].lineNumbers.block.end;
+    }
+    const leadingNonRelevantLines = lines.slice(
+      startWhitespace,
+      update.lineNumbers.block.start
+    );
+    sections.push(leadingNonRelevantLines);
+
+    const providerBlockLines = lines.slice(
+      update.lineNumbers.block.start,
+      update.lineNumbers.block.end
+    );
+    const newProviderBlockLines: string[] = [];
+    let hashLinePrefix = '';
+    let hashLineSuffix = '';
+    providerBlockLines.forEach((providerBlockLine, providerBlockIndex) => {
+      const versionLine = providerBlockLine.replace(
+        versionLineRegex,
+        `$1${update.newVersion}$3`
+      );
+      if (versionLine !== providerBlockLine) {
+        newProviderBlockLines.push(versionLine);
+        return;
+      }
+
+      const constraintLine = providerBlockLine.replace(
+        constraintLineRegex,
+        `$1${update.newConstraint}$3`
+      );
+      if (constraintLine !== providerBlockLine) {
+        newProviderBlockLines.push(constraintLine);
+        return;
+      }
+
+      const hashLineRegexResult = hashLineRegex.exec(providerBlockLine);
+      if (hashLineRegexResult) {
+        // skip hash line but safe the whitespace
+        hashLinePrefix = hashLineRegexResult.groups.prefix;
+        hashLineSuffix = hashLineRegexResult.groups.suffix;
+        return;
+      }
+      newProviderBlockLines.push(providerBlockLine);
+    });
+    const hashesWithWhitespace = update.newHashes.map(
+      (value) => `${hashLinePrefix}${value}${hashLineSuffix}`
+    );
+    newProviderBlockLines.splice(
+      update.lineNumbers.hashes.start,
+      0,
+      ...hashesWithWhitespace
+    );
+    sections.push(newProviderBlockLines);
+  });
+
+  const trailingNotUpdatedLines = lines.slice(
+    updates[updates.length - 1].lineNumbers.block.end
+  );
+  sections.push(trailingNotUpdatedLines);
+
+  const newLines = sections.reduce((previousValue, currentValue) =>
+    previousValue.concat(currentValue)
+  );
+  const newContent = newLines.join('\n');
+
+  const result: UpdateArtifactsResult = {
+    file: {
+      name: lockFile,
+      contents: newContent,
+    },
+  };
+  return new Promise<UpdateArtifactsResult[]>((resolve) => resolve([result]));
```

Comment:
```suggestion
  return Promise.resolve([result]);
```
Simplify. Better we make this function sync.

Reply:
Is it really 'better' or only simpler. It's a genuine question.

Reply:
We prefer simpler code, because it's easier to read and understand. 😏

Reply:
K, I thought maybe there are other implications to this in sync here